### PR TITLE
Enable comparison between Decimal and Float

### LINKED ISSUE #22626: No operation greater between Decimal(9, 2) and Float64
**Describe the bug**
It's not possible to compare Decimal and Float values. But actually both of them point to some specific value on real number line, so we can compare and that wouldn't produce any precision loss.

**How to reproduce**
Clickhouse 21.5
```
SELECT toDecimal32('11.00', 2) > 1.

Query id: f2c1cba1-748e-42ea-83ac-e117b542d576


0 rows in set. Elapsed: 0.002 sec.

Received exception from server (version 21.5.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: No operation greater between Decimal(9, 2) and Float64: While processing toDecimal32('11.00', 2) > 1..
```

Loosely related https://github.com/ClickHouse/ClickHouse/issues/5690

### PR DESCRIPTION
**Pull request by ucasfl**
Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Enable comparison between `Decimal` and `Float`. Closes #22626.

### CODE REVIEW

**Comment by ucasfl** [comment]
@alexey-milovidov Hi, can you help review this? I have no idea about the failed tests.

**Comment by qoega** [comment]
You have to look at this one as it it probably related.
https://s3.amazonaws.com/clickhouse-test-reports/31966/d0c5a887a35a2b89ff8adff600a44ba797235792/stateless_tests__address__actions_.html
993 test is not related - it triggers rare bugs

**Comment by vdimir** [comment]
@mergifyio update

**Comment by ucasfl** [comment]
> @Mergifyio update

Why we need add the label everytime to run tests, does it is a problem of the ci?

**Comment by ucasfl** [comment]
Current implementation is totally wrong, I will re-implement it later. I have some thoughts that I can refer to the comparison between float and int, but not very clear.

**Comment by vdimir** [comment]
> Why we need add the label everytime to run tests, does it is a problem of the ci?

Fixed in https://github.com/ClickHouse/ClickHouse/pull/32696

> Current implementation is totally wrong, I will re-implement it later. I have some thoughts that I can refer to the comparison between float and int, but not very clear.

Let's convert to draft / close ?

File: src/Core/DecimalFloatComparison.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        /// TODO need to implement comparison
+        if (a)
+            return -1;
+        if (b)
+            return 0;
+        if (scale)
+            return 1;
+        return 0;
+    }
```
Comment (ucasfl) [inline src/Core/DecimalFloatComparison.h:20]: Now, I left this function to implement, could you first take a look, does current design is ok? @CurtizJ

Comment (ucasfl) [reply src/Core/DecimalFloatComparison.h:20]: However, I still have no clear idea about the implementation, I checked some systems, in MySQL and Spark, they always convert decimal into float before comparing.

**Comment by ucasfl** [comment]
```
2021-12-24 10:16:54 01460_mark_inclusion_search_crash:                                      [ FAIL ] 0.31 sec. - return code: 158
2021-12-24 10:16:54 [12c190657c8b] 2021.12.24 10:16:54.660055 [ 32172 ] {28ebe7d9-5955-45bc-a89e-a2dfdf53fc8f} <Error> executeQuery: Code: 158. DB::Exception: Limit for rows (controlled by 'max_rows_to_read' setting) exceeded, max rows: 5.00, current rows: 6.00. (TOO_MANY_ROWS) (version 21.13.1.1238) (from [::1]:35030) (comment: 01460_mark_inclusion_search_crash.sql) (in query: SELECT toUInt32(x), y, z FROM pk WHERE (x >= toDateTime(100000))
```
Tests failed seems not related.

**Comment by ucasfl** [comment]
@Mergifyio update.

**Comment by ucasfl** [comment]
@CurtizJ Hi, could you help review this? It's already ready for review.

**Comment by ucasfl** [comment]
@Mergifyio update.

File: src/Functions/FunctionsComparison.h [hunk 2]
```diff
@@ -1232,11 +1235,20 @@ class FunctionComparison : public IFunction
...
             else
             {
-                // compare
+                /// Check does another data type is comparable to Decimal, includes Int and Float.
                 if (!allowDecimalComparison(left_type, right_type) && !date_and_datetime)
                     throw Exception(
                         "No operation " + getName() + " between " + left_type->getName() + " and " + right_type->getName(),
                         ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+                /// When Decimal comparing to Float32/64, we convert both of them into Float64.
+                /// Other systems like MySQL and Spark also do as this.
```
Comment (CurtizJ) [inline src/Functions/FunctionsComparison.h:1244]: Will it work correctly for `Decimal128` and `Decimal256`?

File: src/Functions/FunctionsComparison.h [hunk 3]
```diff
@@ -1232,11 +1235,20 @@ class FunctionComparison : public IFunction
...
+                if (left_is_float || right_is_float)
+                {
+                    const auto converted_type = std::make_shared<DataTypeFloat64>();
+                    ColumnPtr c0_converted = castColumn(col_with_type_and_name_left, converted_type);
+                    ColumnPtr c1_converted = castColumn(col_with_type_and_name_right, converted_type);
+                    return executeGenericIdenticalTypes(c0_converted.get(), c1_converted.get());
```
Comment (CurtizJ) [inline src/Functions/FunctionsComparison.h:1250]: Maybe use more efficient implementation than generic? Something like `executeNum...`. Or for simplicity you can call `executeImpl` once again with new arguments.

File: src/Functions/FunctionsComparison.h (hunk 2 above)
Comment (ucasfl) [reply src/Functions/FunctionsComparison.h:1244]: Yes, it also work well for `Decimal128` and `Decimal256`.

**Comment by ucasfl** [comment]
@Mergifyio update.

**Review by CurtizJ — APPROVED** [review]

**Comment by CurtizJ** [comment]
Test `02156_storage_merge_prewhere` already fixed in master.

**Pull request merged**